File: apps/divsoup/lib/divdump/analyzer/achievements/div_soup/bronze.ex
defmodule Divsoup.Achievement.DivSoupBronze do
  alias Divsoup.Achievement

  @behaviour Divsoup.AchievementRule

  @impl true
  def evaluate(html_tree) do
    ratio = Divsoup.Util.DivSoup.get_div_ratio(html_tree)
    ratio > 0.25
  end

  @impl true
  def achievement() do
    %Achievement{
      hierarchy: :bronze,
      title: "Div Soup",
      description: "More than 25% of the HTML elements in the page are divs"
    }
  end
end

File: src/achievements/classy/bronze.ts
import type { AchievementRule } from "../types";
import { getClassRatio } from "../utils";

export const rule: AchievementRule = {
  id: "classy.bronze",
  title: "Classy",
  group: "classy",
  description:
    "HTML <code>class</code> attributes make up more than <strong>10%</strong> of the page's size",
  hierarchy: "bronze",
  evaluate: ({ doc, rawHtml }) => getClassRatio(doc, rawHtml) > 0.1,
};
